acceptance (mostly) without docker

tbg opened the pull request:
See https://github.com/cockroachdb/cockroach/issues/17464.

```
$ time go test -v -tags acceptance ./pkg/acceptance/ -run TestBuildInfo
--- PASS: TestBuildInfo (2.11s) PASS ok
github.com/cockroachdb/cockroach/pkg/acceptance	2.245s

Today's acceptance tests:
--- PASS: TestBuildInfo (18.84s) PASS ok
github.com/cockroachdb/cockroach/pkg/acceptance	18.960s
```

There isn't really quite that much to see in this PR just yet, and I'm sure
a lot of the other tests break in random ways, but I don't really have the
time to get into this refactor right now. I do have preliminary points
worth discussing though:

1. To do this seriously, I'll have to make a bunch of changes to the
  `localcluster`. I'm planning to rename, move, and generalize it
appropriately.
1. There are quite a few tests that will continue to require
Docker: the
  language client tests (java, php, ...). These typically don't use a full
  Docker cluster but simply run their client code in a container plus they
  need *some* CockroachDB to talk to.
  I'll basically strip down the existing Docker stuff to accommodate just
  that, and make these tests a separate package. As a first pass, I'll
  probably just move all the tests that won't need Docker when I'm done
  somewhere else.
1. For the remaining tests, I'll set up a cluster *once*
and wait for it to
  rebalance, then save the directories. That should make things a lot
faster.
1. We can then discuss running a few of those tests in parallel,
though I
  don't think we'll want to go there.
1. We will want more "reference
tests" that do migrations and such. My initial
  plan here is to make these tests download specific binaries (and cache
them)
  to avoid the awkward postgres-test inspired "shellscript inside Docker"
  crutch we have today, and run them via the `BareCluster` machinery.
1. I'm not planning to change anything about how terraform-enabled tests work.

Please let me know if you have any concerns.

cc @tamird @petermattis

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17686)
<!-- Reviewable:end -->

petermattis commented:
No concerns. Glad to see this happening.

bdarnell commented:
In addition to the language tests that use docker to get their dependencies, a lot of the acceptance tests make network-related assumptions that are only guaranteed to be true in docker. I think all the tests currently assume that port `:26257` is available, which we don't want to rely on (I shouldn't have to take down my local cockroach server to run the acceptance tests). Any tests that are moved out of docker should bind to `127.0.0.1:0` and use `--listening-url-file` instead of using a hard-coded port.

tschottdorf commented:
This now works! If CI is green here, it means that some of our acceptance tests are now running without Docker.

For posterity I'm not editing the original message. I used this to de-flake the decommissioning test, and have run it many a times locally without problems and instead with much delight.

Soo, I know it's close to freeze, but how do we feel about landing this this week? It'll be much easier to write mixed cluster version tests using `BareCluster` than it would be using Docker. That's pretty much all I have left to finish up the version migration work, and it would help me out there a lot. I haven't seen any flakiness locally, though I wouldn't be surprised if "life found a way", so there might be a little bit of CI fixing post-merge.

Ignore the `crdb_internal` part of the patch, that's https://github.com/cockroachdb/cockroach/pull/17762.

tschottdorf commented:
Oh, and.. don't bother reviewing individual commits. They're the opposite of clean and I haven't squashed since I'm still nursing a hope of splitting it into various commits when cleaning up.

petermattis commented on pkg/acceptance/barecluster/localcluster.go:
```diff
@@ -138,61 +187,59 @@ func (c *Cluster) Close() {
+		fmt.Sprintf("--http-port=%d", cfg.HTTPPort),
+		fmt.Sprintf("--store=%s", cfg.DataDir),
+		fmt.Sprintf("--listening-url-file=%s", node.listeningURLFile()),
+		fmt.Sprintf("--pid-file=%s", node.pidFile()),
```
It'll probably become obvious later, but why do you need `--pid-file`. Since we're exec'ing the process we'll know the pid, right?

petermattis commented on pkg/acceptance/barecluster/localcluster.go:
```diff
@@ -42,56 +47,83 @@ import (
 
+// A ClusterConfig holds the configuration for a Cluster.
+type ClusterConfig struct {
+	Ephemeral   bool
```
These fields deserve some documentation, but I'm sure you know that already.

petermattis commented on pkg/acceptance/barecluster/localcluster.go:
```diff
@@ -42,56 +47,83 @@ import (
+	DataDir             string
+	Addr                string
+	ExtraArgs, ExtraEnv []string
+	RPCPort, HTTPPort   int // zero for auto-assign
```
For tests, do we ever want these to be non-zero? Ah, I see you're using these for backwards compatibility with `zerosum` and `allocsim`. Carry on.

petermattis commented on pkg/acceptance/barecluster/localcluster.go:
```diff
@@ -116,16 +142,39 @@ func (c *Cluster) Start(
+			// TODO(tschottdorf): this limits being able to set up a cluster
+			// first and clone it for each test, because all ports change so the
+			// cluster won't come together. Luckily, it takes only ~2 seconds
+			// from zero to a replicated 4 node cluster.
```
Is there anything actionable about this `TODO`? I'm not seeing a way to avoid having to know the RPCPort of the first node. Perhaps `s/TODO(tschottdorf)/Note:/g`.

petermattis commented on pkg/acceptance/barecluster/localcluster.go:
```diff
@@ -425,15 +513,139 @@ func (n *Node) Start() {
+			continue
+		}
+
+		pid := strings.TrimSpace(string(pidBytes))
```
Is this different from `n.cmd.Process.Pid`? If yes, it deserves a comment. Seems like they will be the same unless `--background` is specified which I don't think you're using, but perhaps I missed that.

petermattis reviewed (approved):
LGTM

I'm sure you're aware there are various unfinished doc comments and fields that deserve commenting. The term "Bare" is slightly non-descriptive for me. Perhaps "Native" or even "Local". The latter would give "local", "remote" and "docker" clusters which seems reasonable.

tamird commented:
Reviewed 6 of 8 files at r1, 2 of 9 files at r2, 5 of 19 files at r4, 1 of 1 files at r5, 2 of 2 files at r10, 1 of 5 files at r13, 20 of 27 files at r14.
Review status: 20 of 25 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/.gitignore, line 4 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFS5qO_FmZtf_tmPiy:-KsFS5qPojD7X8GKTHtP:bilwkas) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/.gitignore#L4)):*
> ```
> # for reasoning and alternatives).
> .localcluster.certs.*
> .barecluster*
> ```

should the rename to local have been propagated to here?

---

*[pkg/acceptance/decommission_test.go, line 53 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRZVjcbxpgeREJqmT:-KsFRZVkOCrcBvDX64CI:b-g278wo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/decommission_test.go#L53)):*
> ```Go
> 		args := append([]string{"node", verbs[0], strconv.Itoa(int(targetNode))}, verbs[1:]...)
> 		o, _, err := c.ExecCLI(ctx, runNode, args)
> 		if testutils.IsError(err, "ambiguous") {
> ```

intentional? perhaps a comment.

---

*[pkg/acceptance/decommission_test.go, line 164 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRdgysGPvatMP3YlC:-KsFRdgysGPvatMP3YlD:blpl2if) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/decommission_test.go#L164)):*
> ```Go
> 			// TODO(tschottdorf): grep the process output for the string announcing success?
> 			t.Logf("ignoring error on quit --decommission: %s", err)
> 			log.Warningf(ctx, "ignoring error on quit --decommission: %s", err)
> ```

did you mean to have both?

---

*[pkg/acceptance/gossip_peerings_test.go, line 94 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRlxfIP2WPE9XdAEP:-KsFRlxfIP2WPE9XdAEQ:b-7ntrut) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/gossip_peerings_test.go#L94)):*
> ```Go
> 	// cluster leads to no node knowing where to reach any other node. To change
> 	// this, we could introduce an RPC we can send to the cluster that gives it
> 	// "hints" for its --join list after it has started.
> ```

could we remember the ports the nodes were first started on and reuse them?

---

*[pkg/acceptance/util.go, line 656 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFT8cdxnbixSb_sJOx:-KsFT8cdxnbixSb_sJOy:bd09qiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/util.go#L656)):*
> ```Go
> 		}
> 		hostConfig := container.HostConfig{NetworkMode: "host"}
> 		if err := l.OneShot(
> ```

this is the only assignment to `err` here; you can just `var err error` outside to avoid the awkward name, I think.

---

*[pkg/acceptance/util.go, line 661 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFTEY-FitGv4GT71tt:-KsFTEY-FitGv4GT71tu:b-kvdn76) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/util.go#L661)):*
> ```Go
> 			rErr = err
> 		}
> 		// Clean up the log files if the run was successful.
> ```

this is now unconditional, yeah?

---

*[pkg/cli/node.go, line 358 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFSUw8kgj7t3K5-mOv:-KsFSUw8kgj7t3K5-mOw:b-ujy9iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/cli/node.go#L358)):*
> ```Go
> 		}
> 	}
> 	panic("unreachable!")
> ```

what was wrong with the old code? looks equivalent.

---

*[pkg/server/server.go, line 390 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFSmrDVUCsXgA4YdBJ:-KsFSmrEd9a9x_qHGuAN:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/server/server.go#L390)):*
> ```Go
> 	}
> 
> 	// FINDME(tschottdorf)
> ```

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 20 of 25 files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[pkg/cmd/zerosum/main.go, line 34 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFUBY_idJL2Fn8MFyl:-KsFUBY_idJL2Fn8MFym:b-5wn5bn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/cmd/zerosum/main.go#L34)):*
> ```Go
> 
> 	"github.com/cockroachdb/cockroach-go/crdb"
> 	barecluster "github.com/cockroachdb/cockroach/pkg/acceptance/localcluster"
> ```

let's rename this back; it should reduce the diff somewhat as well. ditto in allocsim

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 20 of 25 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/.gitignore, line 4 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFS5qO_FmZtf_tmPiy:-KsFVCzyRAoXSKnQcsoy:b-745nwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/.gitignore#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should the rename to local have been propagated to here?
</blockquote></details>

Absolutely, done.

---

*[pkg/acceptance/decommission_test.go, line 53 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRZVjcbxpgeREJqmT:-KsFVZ1vCn4qXJH68zdO:b-bizqh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/decommission_test.go#L53)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

intentional? perhaps a comment.
</blockquote></details>

Added comment.

---

*[pkg/acceptance/decommission_test.go, line 164 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRdgysGPvatMP3YlC:-KsFVcumtuS4YuGJfe_O:b2r0uju) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/decommission_test.go#L164)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

did you mean to have both?
</blockquote></details>

I think I really want the warning. Removed the other one.

---

*[pkg/acceptance/gossip_peerings_test.go, line 94 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRlxfIP2WPE9XdAEP:-KsFVgLzMI_B-LknemBU:bspz6gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/gossip_peerings_test.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

could we remember the ports the nodes were first started on and reuse them?
</blockquote></details>

Yes, there's a TODO about this somewhere. I'll do more refactors in that direction when this merges (I also have to bring back `Assert()`, and get more tests over to the new system).

---

*[pkg/acceptance/util.go, line 656 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFT8cdxnbixSb_sJOx:-KsFW9AX6pCCAOMi2GRR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/util.go#L656)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is the only assignment to `err` here; you can just `var err error` outside to avoid the awkward name, I think.
</blockquote></details>

Done.

---

*[pkg/acceptance/util.go, line 661 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFTEY-FitGv4GT71tt:-KsFW5pcJbLvXRwKhEv5:b9fwkpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/util.go#L661)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is now unconditional, yeah?
</blockquote></details>

That was unintentional. Restored the old behavior.

---

*[pkg/cli/node.go, line 358 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFSUw8kgj7t3K5-mOv:-KsFWc61DC1eCZdpBZrR:bom1s9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/cli/node.go#L358)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

what was wrong with the old code? looks equivalent.
</blockquote></details>

It's equivalent, but the old code was vulnerable to someone changing the retry options to use a MaxRetries, and then no error would be returned when there should be one. However, that certainly didn't get reflected in the code (not sure what I was doing there). Fixed up.

---

*[pkg/cmd/zerosum/main.go, line 34 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFUBY_idJL2Fn8MFyl:-KsFWx36TgoACqFlY4Ny:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/cmd/zerosum/main.go#L34)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

let's rename this back; it should reduce the diff somewhat as well. ditto in allocsim
</blockquote></details>

Done.

---

*[pkg/server/server.go, line 390 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFSmrDVUCsXgA4YdBJ:-KsFWzjkuphPsn8oQwQX:b-w0jof5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/server/server.go#L390)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

?
</blockquote></details>

As spurious as it gets, removed.

---

*[pkg/acceptance/barecluster/localcluster.go, line 70 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-Ks6NI9DGMQSAzZW9n4k-r13-70:-KsFFZfz9_GNJyrWFYYQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98e34897fd10eb1125514a04ce00517e5b478793/pkg/acceptance/barecluster/localcluster.go#L70)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

These fields deserve some documentation, but I'm sure you know that already.
</blockquote></details>

Done.

---

*[pkg/acceptance/barecluster/localcluster.go, line 169 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-Ks6NI9DGMQSAzZW9n4k-r13-169:-KsFJ45y3xmvSCkPbB2n:busqs4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/98e34897fd10eb1125514a04ce00517e5b478793/pkg/acceptance/barecluster/localcluster.go#L169)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is there anything actionable about this `TODO`? I'm not seeing a way to avoid having to know the RPCPort of the first node. Perhaps `s/TODO(tschottdorf)/Note:/g`.
</blockquote></details>

We could make it work: Add an RPC that basically makes a gossip endpoint known to the node when it's already running. Start each node with `--join=whatever`. Call the endpoint when you've collected the actual listening ports. Or, run a small service on the side that the nodes talk to and know how to get the address of, ...

Not worth it obviously, so I've changed the wording as you suggested.

---

*[pkg/acceptance/barecluster/localcluster.go, line 229 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-Ks6NI9DGMQSAzZW9n4k-r13-229:-KsFJSPjXxzK_WMQ2K1j:btify13) ([raw file](https://github.com/cockroachdb/cockroach/blob/98e34897fd10eb1125514a04ce00517e5b478793/pkg/acceptance/barecluster/localcluster.go#L229)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It'll probably become obvious later, but why do you need `--pid-file`. Since we're exec'ing the process we'll know the pid, right?
</blockquote></details>

Honestly, I'm not quite sure any more. Removed.

---

*[pkg/acceptance/barecluster/localcluster.go, line 585 at r13](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-Ks6NI9DGMQSAzZW9n4k-r13-585:-KsFK2lPNZJh5i0hC2Ss:b-ljrwwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/98e34897fd10eb1125514a04ce00517e5b478793/pkg/acceptance/barecluster/localcluster.go#L585)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this different from `n.cmd.Process.Pid`? If yes, it deserves a comment. Seems like they will be the same unless `--background` is specified which I don't think you're using, but perhaps I missed that.
</blockquote></details>

No, you didn't miss anything. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tamird commented:
Skimmed localcluster.go, LGTM. Could do a thorough pass after current comments, if you'd like.

---

Reviewed 2 of 27 files at r14, 7 of 8 files at r15.
Review status: 24 of 25 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[pkg/acceptance/gossip_peerings_test.go, line 94 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRlxfIP2WPE9XdAEP:-KsFdajLoEmwf5EkiKGz:b-4umvo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/gossip_peerings_test.go#L94)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yes, there's a TODO about this somewhere. I'll do more refactors in that direction when this merges (I also have to bring back `Assert()`, and get more tests over to the new system).
</blockquote></details>

Ack. Might be worthwhile to add that idea here for future readers.

---

*[pkg/acceptance/util.go, line 495 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFc20EpMiexBhUfGv9:-KsFc20EpMiexBhUfGvA:bjuqsvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L495)):*
> ```Go
> 			clusterCfg := localcluster.ClusterConfig{
> 				Ephemeral: true,
> 				Binary:    "../../cockroach",
> ```

somewhat unfortunate that this is relative. you could use go/build to make it absolute (and I think that's already done in one of the flags in the acceptance package).

---

*[pkg/acceptance/util.go, line 514 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFbhEVoUo81ExdjtI-:-KsFbhEVoUo81ExdjtI0:b-azqtkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L514)):*
> ```Go
> 
> 		default:
> 			t.Fatalf("unable to run in mode %q, use either RunLocal, RunDocker, or RunFarmer", parts[1])
> ```

s/parts[1]/runMode/ ... or `parts[1:]`

---

*[pkg/acceptance/localcluster/cluster.go, line 30 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcNTj7_B90L5vVlE6:-KsFcNTj7_B90L5vVlE7:b-u0svr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/cluster.go#L30)):*
> ```Go
> 
> // LocalCluster implements cluster.Cluster.
> type LocalCluster struct {
> ```

i think this file and `localcluster.go` are named in reverse =/

---

*[pkg/acceptance/localcluster/cluster.go, line 76 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcZadXT-1j7S1976t:-KsFcZadXT-1j7S1976u:brrffnx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/cluster.go#L76)):*
> ```Go
> // ExecCLI implements cluster.Cluster.
> func (b *LocalCluster) ExecCLI(ctx context.Context, i int, cmd []string) (string, string, error) {
> 	cmd = append([]string{"../../cockroach"}, cmd...)
> ```

also duplicates the binary location.

---

*[pkg/acceptance/localcluster/localcluster.go, line 68 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcp4dDG-83xXNJvCp:-KsFcp4dDG-83xXNJvCq:bhypns3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L68)):*
> ```Go
> type ClusterConfig struct {
> 	Ephemeral   bool               // when true, wipe DataDir on Close()
> 	Binary      string             // path to cockroach, default ../../cockroach
> ```

path duplicated once more

---

*[pkg/acceptance/localcluster/localcluster.go, line 92 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFaq-UHVsA2UHgOvGf:-KsFaq-UHVsA2UHgOvGg:b-7wko1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L92)):*
> ```Go
> // with odd ports starting at 26257 for the RPC endpoint, and even points for
> // the ui.
> func MakePerNodeFixedPortsCfg(numNodes int) map[int]NodeConfig {
> ```

why is this a map rather than a slice? is the intention that elements can be removed?

--

ah, perhaps so that you can retrieve defaults? feel free to ignore related comments.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsFdG42Lvn57K5FcLQF:b-sgy3ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
> ```Go
> func (n *Node) Start() {
> 	var wg sync.WaitGroup
> 	n.StartAsync(&wg)
> ```

startAsync has only one other caller; it might be cleaner to return (func(), error) from Start so that the caller can decide if they want to wait (that would avoid this awkward plumbing of waitgroups).

---

*[pkg/acceptance/localcluster/localcluster.go, line 530 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFe2pkp90f14WfAVyi:-KsFe2pkp90f14WfAVyj:b-t0jqcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L530)):*
> ```Go
> 	conn, err := gosql.Open("postgres", url)
> 	if err != nil {
> 		log.Fatal(context.Background(), err)
> ```

there are 11 log.Fatals in this file. it'd be nice to reduce that number =/

---

*[pkg/cli/node.go, line 313 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsF_rQpgXoQFJtyFeO5:-KsF_rQq_gmQOaJo_0ry:b-1va45l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/cli/node.go#L313)):*
> ```Go
> 	}
> 	minReplicaCount := int64(math.MaxInt64)
> ```

spurious newline

---

*[pkg/cmd/allocsim/main.go, line 440 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFaKKfFQpKOBpo2NPD:-KsFaKKfFQpKOBpo2NPE:b3bzk3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/cmd/allocsim/main.go#L440)):*
> ```Go
> 	if separateAddrs {
> 		for i := range perNodeCfg {
> 			s := perNodeCfg[i]
> ```

hm, why not `perNodeCfg[i].Addr = fmt.Sprintf(...)`?

--

heh, looks like it's a map, but i don't grok why. left a comment there.

---

*[pkg/cmd/allocsim/main.go, line 455 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFaZ-fIg0h9utQxNlY:-KsFaZ-fIg0h9utQxNlZ:bfx8zw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/cmd/allocsim/main.go#L455)):*
> ```Go
> 		for _, locality := range config.Localities {
> 			for i := 0; i < locality.NumNodes; i++ {
> 				s := perNodeCfg[nodeIdx] // avoid map assignment problems
> ```

ditto, could be `s := &perNodeCfg[nodeIdx]` provided perNodeCfg changes to a slice

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-:-KsFe8O_AQEbPGrZa0Ql:b4tbg3d)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 9 files at r2, 4 of 19 files at r4, 2 of 2 files at r10, 16 of 27 files at r14, 8 of 8 files at r15.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/debug_remote_test.go, line 38 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFlbg3Yzox05HEf3SR:-KsFlbg3Yzox05HEf3SS:bhgoi46) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/debug_remote_test.go#L38)):*
> ```Go
> 
> func testDebugRemote(t *testing.T) {
> 	SkipUnlessLocal(t)
> ```

"Local" in `SkipUnlessLocal` looks like it means something different from `RunLocal`

---

*[pkg/acceptance/gossip_peerings_test.go, line 94 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRlxfIP2WPE9XdAEP:-KsFm3uGsSi-X1Ui74Zl:bl5ryd9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/gossip_peerings_test.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Ack. Might be worthwhile to add that idea here for future readers.
</blockquote></details>

If we restarted the nodes one at a time, we could get the first node's new port and pass it as `--join` to all the subsequent ones. 

The port that we used before is not guaranteed to remain available (not really an issue for us, but public CI servers churn through ephemeral ports at a ridiculous rate and you're likely to see errors whenever you bind to a port that's not guaranteed to be yours)

---

*[pkg/acceptance/localcluster/localcluster.go, line 547 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFomzzfRADljqyQkPS:-KsFomzzfRADljqyQkPT:bkdrsqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L547)):*
> ```Go
> 		Multiplier:     2,
> 	}
> 	for r := retry.Start(opts); r.Next(); {
> ```

If you made the listeningURLFile a named pipe with syscall.Mkfifo, you wouldn't need a retry loop here (see the way this is used in the tcl tests). May not be worth the portability headaches, though, if we want to get these tests running on windows later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 17 of 25 files reviewed at latest revision, 17 unresolved discussions.

---

*[pkg/acceptance/debug_remote_test.go, line 38 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFlbg3Yzox05HEf3SR:-KsFzFMyAPohSl0pT0zd:bmoukml) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/debug_remote_test.go#L38)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

"Local" in `SkipUnlessLocal` looks like it means something different from `RunLocal`
</blockquote></details>

Removed SkipUnlessLocal. I believe it's not used in practice: when running `make acceptance` it's always true, and when running the farmer tests we always target an individual test anyway.
There's more refactoring to do here, but I'll (try to) avoid breaking the surface used by CI for now.

---

*[pkg/acceptance/gossip_peerings_test.go, line 94 at r14](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFRlxfIP2WPE9XdAEP:-KsFzXP6xxd8wvQh_KNC:bxj3ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/d856c236ccf930174fc9430d5b15d905c0dc0fc2/pkg/acceptance/gossip_peerings_test.go#L94)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If we restarted the nodes one at a time, we could get the first node's new port and pass it as `--join` to all the subsequent ones. 

The port that we used before is not guaranteed to remain available (not really an issue for us, but public CI servers churn through ephemeral ports at a ridiculous rate and you're likely to see errors whenever you bind to a port that's not guaranteed to be yours)
</blockquote></details>

That's a good point, Ben. Unfortunately it's not always as easy as "restart all nodes". Some tests are chaos-y and I've seen the cluster never come back together. Might be tickling a few bugs or just be expected, I'll have to look more. Prefer not to get sidetracked in this PR though, updated the comment though.

---

*[pkg/acceptance/util.go, line 495 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFc20EpMiexBhUfGv9:-KsG-hj1ZteljLlsO02T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L495)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

somewhat unfortunate that this is relative. you could use go/build to make it absolute (and I think that's already done in one of the flags in the acceptance package).
</blockquote></details>

Done.

---

*[pkg/acceptance/util.go, line 514 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFbhEVoUo81ExdjtI-:-KsG-4WJUPowP3i_ETsq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L514)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/parts[1]/runMode/ ... or `parts[1:]`
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/cluster.go, line 30 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcNTj7_B90L5vVlE6:-KsG-KzJPDsPhzIDYymh:b-flual9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/cluster.go#L30)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think this file and `localcluster.go` are named in reverse =/
</blockquote></details>

Ew. Done.

---

*[pkg/acceptance/localcluster/cluster.go, line 76 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcZadXT-1j7S1976t:-KsG-WyCfPI4kI6DjlKx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/cluster.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

also duplicates the binary location.
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/localcluster.go, line 68 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFcp4dDG-83xXNJvCp:-KsG32_HquVIHs0MXhrI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L68)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

path duplicated once more
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/localcluster.go, line 92 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFaq-UHVsA2UHgOvGf:-KsG34OVDKCLOpjqwRhL:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L92)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this a map rather than a slice? is the intention that elements can be removed?

--

ah, perhaps so that you can retrieve defaults? feel free to ignore related comments.
</blockquote></details>

Yep.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsG3OOldRnXnDmvIUtt:b4ltu0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

startAsync has only one other caller; it might be cleaner to return (func(), error) from Start so that the caller can decide if they want to wait (that would avoid this awkward plumbing of waitgroups).
</blockquote></details>

I prefer to keep the signature of `Start()` as people expect it.

---

*[pkg/acceptance/localcluster/localcluster.go, line 530 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFe2pkp90f14WfAVyi:-KsG3TGvE_plye-lzu7O:b-4rkdoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L530)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

there are 11 log.Fatals in this file. it'd be nice to reduce that number =/
</blockquote></details>

Agreed, but it's not something I'm going to pursue in this PR.

---

*[pkg/acceptance/localcluster/localcluster.go, line 547 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFomzzfRADljqyQkPS:-KsG3y4kERvI0pG4drSp:bmh6lc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L547)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

If you made the listeningURLFile a named pipe with syscall.Mkfifo, you wouldn't need a retry loop here (see the way this is used in the tcl tests). May not be worth the portability headaches, though, if we want to get these tests running on windows later.
</blockquote></details>

I'll keep this low-tech for now. Would be nice to run acceptance on windows.

---

*[pkg/cli/node.go, line 313 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsF_rQpgXoQFJtyFeO5:-KsG43jfcUbCJSaSHg5k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/cli/node.go#L313)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

spurious newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 27 files at r14, 7 of 8 files at r16.
Review status: 24 of 25 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/util.go, line 495 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFc20EpMiexBhUfGv9:-KsGlxR2uRP5zFAYbYRn:b-cv3a70) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L495)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks unchanged.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsGmMPkjofZgR59mphG:b-kaxckk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I prefer to keep the signature of `Start()` as people expect it.
</blockquote></details>

OK. You can make StartAsync (or whatever new name it might want) do what I suggested, and just have Start call the returned function. Up to you, though I did find it difficult to trace waitgroup through the current implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 24 of 25 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/debug_remote_test.go, line 33 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ3jt0EYK5qN-vmYqy:-KsJ3jt0EYK5qN-vmYqz:bod64f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/debug_remote_test.go#L33)):*
> ```Go
> 	s := log.Scope(t)
> 	defer s.Close(t)
> 	// TODO(tschottdorf): with some love, this might run under RunLocal.
> ```

Might be challenging to get it stable under `RunLocal` as this requires specific networking setup.

---

*[pkg/acceptance/localcluster/cluster.go, line 450 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ4ousxcxgWcvjfG-Q:-KsJ4ousxcxgWcvjfG-R:bo8vyg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/localcluster/cluster.go#L450)):*
> ```Go
> }
> 
> // StartAsync .
> ```

Incomplete comment here.

---

*[pkg/acceptance/localcluster/cluster.go, line 515 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ4yuyDaGquMduVlIw:-KsJ4yuyDaGquMduVlIx:b-sr6lhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/localcluster/cluster.go#L515)):*
> ```Go
> 		}
> 
> 		log.Infof(ctx, "Process %d: %s", cmd.Process.Pid, cmd.ProcessState)
> ```

`s/Process/process/g` for consistency with the other log messages.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-:-KsJ5BM8yA0lQSOGTskk:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 22 of 25 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/acceptance/debug_remote_test.go, line 33 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ3jt0EYK5qN-vmYqy:-KsJniq9MmpJt9S6ocE2:b-4rza8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/debug_remote_test.go#L33)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Might be challenging to get it stable under `RunLocal` as this requires specific networking setup.
</blockquote></details>

That's a good point, updated the comment.

---

*[pkg/acceptance/util.go, line 495 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFc20EpMiexBhUfGv9:-KsJmZJjJU9s8ALHExGF:b2om9d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/util.go#L495)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Looks unchanged.
</blockquote></details>

Apologies, done'd the wrong comment. Fixed.

---

*[pkg/acceptance/localcluster/cluster.go, line 450 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ4ousxcxgWcvjfG-Q:-KsJmqS_2GrwongrKv7i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/localcluster/cluster.go#L450)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Incomplete comment here.
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/cluster.go, line 515 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsJ4yuyDaGquMduVlIw:-KsJmsErKN31L3pa0qA-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d4276b60d78c9f89187196a334dd50b72657e75/pkg/acceptance/localcluster/cluster.go#L515)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/Process/process/g` for consistency with the other log messages.
</blockquote></details>

Done.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsJnPvNIsZNFTLKVoem:bpo8dvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

OK. You can make StartAsync (or whatever new name it might want) do what I suggested, and just have Start call the returned function. Up to you, though I did find it difficult to trace waitgroup through the current implementation.
</blockquote></details>

The WaitGroup is created in `(c *Cluster).Start()`, where we wait for all but the first node in parallel. I don't see how what you're suggesting is better. Will leave as is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r17.
Review status: 24 of 25 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsJuZK3G6AN4rD8HlQG:b20os0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

The WaitGroup is created in `(c *Cluster).Start()`, where we wait for all but the first node in parallel. I don't see how what you're suggesting is better. Will leave as is.
</blockquote></details>

My suggestion results in this code in `(*Cluster).Start`:
```
var wg sync.WaitGroup
for _, n := range c.Nodes {
  fn, err := n.StartAsync()
  if err ...
  go func() { wg.Add(1); defer wg.Done(); fn() }()
}
wg.Wait()
```

I think it's better because it's easy to spot what the wg is for. Again, not blocking you on this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 24 of 25 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsJvQG0RSQ97hG425tZ:bqmfen6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

My suggestion results in this code in `(*Cluster).Start`:
```
var wg sync.WaitGroup
for _, n := range c.Nodes {
  fn, err := n.StartAsync()
  if err ...
  go func() { wg.Add(1); defer wg.Done(); fn() }()
}
wg.Wait()
```

I think it's better because it's easy to spot what the wg is for. Again, not blocking you on this.
</blockquote></details>

Er, the `wg.Add` call should be outside the spawned goroutine, but you get the idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tbg committed:
acceptance: use less Docker

See https://github.com/cockroachdb/cockroach/issues/17464.

Rename our (Docker-based) localcluster to `dockercluster`. Introduce a new
`localcluster` package, which is a revamp of the formerly internal
`./pkg/cmd/internal/localcluster` adapted to the `cluster.Cluster` interface.

Use it for all tests to which it immediately applies. Due to the use of
ephemeral ports, whole-cluster restarts are currently a problem (though we can
mitigate: save the old port numbers and reassign them when a down node
restarts).

Additionally, no attempt was made here to refactor the client test machinery or
the reference tests. This work is slated to happen in the future.

The immediate upshot of this mode is that debugging flaky tests or iterating on
new ones is greatly simplified.

I expect this to make it much more pleasant to introduce serious mixed-version
cluster acceptance + migration tests, which is my last remaining "feature-ish"
item for this release cycle.

Running these tests is simple. Tests encode which type of cluster they were
written for using the `Run{Docker,Local,Terraform}` helpers that work similarly
to `t.Run(...)`, but pick a name for you that signals to our test runner the
correct type of cluster. That is, `make acceptance` will do the right thing.

Since `make acceptance` still prepares the builder and builds a binary, it is
always going to be more appropriate to run `go test` directly when all you want
is a localcluster-enabled test, like so:

```
time go test -v -tags acceptance ./pkg/acceptance/ -run Decom -v -show-logs
```

This commit also unskips TestDecommission (fixed in
https://github.com/cockroachdb/cockroach/pull/17724), but clarifies some errors
that were unhelpful while fixing it.

tbg committed:
storage: avoid spurious liveness update failures

tbg committed:
acceptance-ci: bump limit to 20 minutes

tschottdorf commented:
Review status: 20 of 26 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/localcluster/localcluster.go, line 434 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686#-KsFdG42Lvn57K5FcLQE:-KsV1ITbVlQxbLYNl2Ta:b8c4d2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d90a16eefc5f634ed21216eb254edfcc999b25cf/pkg/acceptance/localcluster/localcluster.go#L434)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Er, the `wg.Add` call should be outside the spawned goroutine, but you get the idea.
</blockquote></details>

Returning a `<- chan error` now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17686)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ran CI a bunch of times, seems to work well - I did bump the CI timeout for acceptance from 10 to 20 minutes (since the decommissioning test isn't exactly fast). Going to merge. 🤞
